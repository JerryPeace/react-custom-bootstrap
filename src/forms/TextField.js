import PropTypes from 'prop-types';
import React from 'react';
import _ from 'lodash';
import {Basic} from './Basic';
import Label from './Label';
import classNames from 'classnames';
import FormGroup from './FormGroup';
import FormControls from './FormControls';
import HelpText from './HelpText';
import LayoutHelper from './LayoutHelper';

class Textfield extends Basic {
    static propTypes = {
        labelText: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
        inputType: PropTypes.oneOf(['text', 'password', 'datetime', 'datetime-local', 'date', 'month', 'time', 'week', 'number', 'email', 'url', 'file']),
        disabled: PropTypes.bool,
        multiline: PropTypes.bool,
        standalone: PropTypes.bool,
        helpText: PropTypes.string,
        helpTextComponent: PropTypes.element,
        helpTextClassName: PropTypes.string,
        onChange: PropTypes.func,
        onValueChange: PropTypes.func,
        onInputChange: PropTypes.func,
        onValidate: PropTypes.func,
        validator: PropTypes.func,
        tooltipProps: PropTypes.object,
        delayValidateTime: PropTypes.number
    };

    static defaultProps = {
        inputType: 'text',
        disabled: false,
        multiline: false,
        bsClass: 'input',
        labelAlign: 'left',
        standalone: true,
        helpText: '',
        labelClassName: 'col-xs-2',
        wrapperClassName: 'col-xs-10',
        helpTextClassName: 'has-error',
        delayValidateTime: 500,
        validator:
            (textField, value) => {
                return true;
            }
    };

    constructor(props) {
        super(props);
        this.oldVal = null;
        this.changeTimeoutID = null;
        this.validate = ::this.validate;
        this.originalVal = ::this._initOriginalVal();
        this._onChange = ::this._onChange;
        this.resultHandler = ::this.resultHandler;
    }

    componentWillUnmount() {
        clearTimeout(this.changeTimeoutID);
    }

    state = {
        helpText: null
    };

    _initOriginalVal() {
        let value;
        if (typeof this.props.value !== 'undefined') {
            value = this.props.value;
        } else if (typeof this.props.defaultValue !== 'undefined') {
            value = this.props.defaultValue;
        } else {
            value = '';
        }

        return value;
    }

    resultHandler(res) {
        const { onValidate } = this.props;
        const validInfo = {
            helpText: null
        };
        const isValid = res === true;

        if (!isValid) {
            validInfo.helpText = typeof res === 'string' ? res : this.props.helpText;
            validInfo.result = false;
        } else {
            validInfo.result = true;
        }

        this.setState({
            helpText: validInfo.helpText
        }, () => {
            onValidate && onValidate(isValid, {
                helpText: validInfo.helpText
            });
        });


        return validInfo;
    };

    getValue() {
        return this._inputEl && this._inputEl.value;
    }

    isValid() {
        const value = this.getValue();
        const result = this.props.validator(this, value); //true, false, string, promise

        return (typeof result === 'object' && _.isFunction(result.then)) ? result : result === true;
    }

    validate() {
        const { validator } = this.props;
        const value = this.getValue();
        const result = validator(this, value); //true, false, string, promise

        if (typeof result === 'object' && _.isFunction(result.then)) {
            return result.then(this.resultHandler);
        } else {
            return this.resultHandler(result);
        }
    }

    markInvalidText(text) {
        this.setState({
            helpText: text
        });
    }

    clearInvalidText() {
        this.markInvalidText(null);
    }

    _onChange(e) {
        let newVal = this.getValue(),
            oldVal = this.oldVal;

        clearTimeout(this.changeTimeoutID);

        this.changeTimeoutID = setTimeout(this.validate, this.props.delayValidateTime);

        if (this.props.onValueChange) {
            console.warn('Deprecated function, please use onInputChange');
            this.props.onValueChange(this, newVal, oldVal, e);
        }

        if (this.props.onInputChange) {
            this.props.onInputChange(newVal, {
                oldValue: oldVal,
                e
            });
        }

        if (this.props.onChange) {
            this.props.onChange(e);
        }

        this.oldVal = newVal;
    }

    getInputElement() {
        return this._inputEl;
    }

    renderInput() {
        const {inputType, className, multiline, helpText, labelText, onValueChange, bsClass, labelAlign, standalone, labelClassName, wrapperClassName, helpTextClassName, validator, bsSize, bsStyle, onChange, onValidate, children, delayValidateTime, ...rest} = this.props;
        const Component = multiline ? 'textarea' : 'input';

        return (
            <Component
                {...rest}
                ref={(el) => this._inputEl = el}
                key={Component}
                type={inputType}
                className={classNames(className)}
                onChange={this._onChange} />
        );
    }

    render() {
        return (
            <LayoutHelper
                {...this.props}
                FieldBase={this.renderInput()}
                helpText={this.props.helpText || this.state.helpText}
                getInputEl={() => this._inputEl} />
        );
    }
}

export default Textfield;
