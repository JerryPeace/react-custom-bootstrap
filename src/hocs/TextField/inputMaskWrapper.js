import PropTypes from 'prop-types';
import React, { Component } from 'react';
import _ from 'lodash';

export default (TextField) => {
    return class inputMaskWrapper extends Component {
        static propTypes = {
            mask: PropTypes.string.isRequired,
            value: PropTypes.string,
            onChange: PropTypes.func,
            onFocus: PropTypes.func,
            onKeyDown: PropTypes.func,
            onPaste: PropTypes.func,
            customRules: PropTypes.object,
            placeholder: PropTypes.string
        };

        constructor(props) {
            super(props);
            this.dataBeforeChange = null;
            this.dataAfterChange = null;
        }

        getMaskRule = () => {
            return {
                '0': {
                    regExp: '\\d',
                    guideChar: '0'
                },
                // for "Activation Code" char
                'X': {
                    regExp: '[a-zA-Z0-9]',
                    guideChar: ' '
                },
                ...this.props.customRules
            };
        }

        getTextRules = () => {
            const maskRules = this.getMaskRule();
            const mask = this.props.mask;
            return _.map(mask, (char) => (maskRules[char] ? maskRules[char] : {
                isTemplate: true,
                guideChar: char
            }));
        }

        getUpdatedData = (preData, nextData) => {
            const {
                value: preValue,
                selectionStart: preSelectionStart,
                isPaste
            } = preData;
            const {
                value: nextValue,
                selectionStart: nextSelectionStart
            } = nextData;

            const textRules = this.getTextRules();

            if (isPaste) {
                // For paste
                const validText = this.addMask(nextValue, textRules);
                return {
                    maskText: _.merge(
                        this.createGuideMask(),
                        validText
                    ).join(''),
                    cusorPos: validText.length
                };
            } else {
                const guideMask = this.createGuideMask();
                const preMaskText = preValue === '' ? guideMask : preValue;
                const {
                    updateStartIdx,
                    updateEndIdx
                } = this.findUpdateRange(preMaskText, textRules, preSelectionStart, nextSelectionStart);

                // Update mask text
                const baseMaskText =  _.map(preMaskText, (char, idx) => {
                    return (idx >= updateStartIdx && idx <= updateEndIdx) ? guideMask[idx] : char;
                });
                const updateRangeMask = _.slice(textRules, updateStartIdx, updateEndIdx + 1);
                const textToBeUpdate = nextValue.substr(updateStartIdx, updateEndIdx - updateStartIdx + 1);
                const updateText = this.addMask(textToBeUpdate, updateRangeMask);
                const maskText = _.concat(
                    _.slice(baseMaskText, 0, updateStartIdx),
                    updateText,
                    _.slice(baseMaskText, updateStartIdx + updateText.length)
                ).join('');

                // Caculate cursor position
                let cusorPos = nextSelectionStart;
                if (nextSelectionStart > preSelectionStart) {
                    if (!updateText.length || preValue === maskText) {
                        cusorPos = preSelectionStart;
                    } else {
                        const maskLength = this.props.mask.length;
                        const nextEditIdx = _.findIndex(
                            maskText,
                            (char, idx) => (!textRules[idx].isTemplate),
                            updateStartIdx + 1
                        );
                        cusorPos = nextEditIdx !== -1 ? nextEditIdx : maskLength;
                    }
                }

                return {
                    maskText: _.concat(
                        _.slice(baseMaskText, 0, updateStartIdx),
                        updateText,
                        _.slice(baseMaskText, updateStartIdx + updateText.length)
                    ).join(''),
                    cusorPos
                };
            }
        }

        findUpdateRange(preMaskText, textRules, preCursorPos, nextCursorPos) {
            const lastIdx = textRules.length - 1;
            if (preCursorPos >= nextCursorPos) {
                // Delete text
                return {
                    updateStartIdx: nextCursorPos,
                    updateEndIdx: lastIdx
                };
            } else {
                // Add text
                const nearbyInvalidIdx = _.findIndex(textRules, (charRule, idx) => {
                    if (charRule.isTemplate || idx < preCursorPos) {
                        return false;
                    } else {
                        const charRegExp = new RegExp(charRule.regExp);
                        return !charRegExp.test(preMaskText[idx]);
                    }
                });
                if (nearbyInvalidIdx === -1) {
                    // Text complete, no need to update
                    return {
                        updateStartIdx: null,
                        updateEndIdx: null
                    };
                } else {
                    const blockIdx = _.findIndex(textRules, (charRule, idx) => {
                        if (charRule.isTemplate || idx < nearbyInvalidIdx) {
                            return false;
                        } else {
                            const charRegExp = new RegExp(charRule.regExp);
                            return charRegExp.test(preMaskText[idx]);
                        }
                    });
                    return {
                        updateStartIdx: preCursorPos,
                        updateEndIdx: blockIdx === -1 ? lastIdx : blockIdx - 1
                    };
                }
            }
        }

        addMask(text, textRules) {
            let textArr = _.toArray(text);
            let maskText = [];
            _.forEach(textRules, (charRule, idx) => {
                if (charRule.isTemplate) {
                    maskText.push(charRule.guideChar);
                } else {
                    const charRegExp = new RegExp(charRule.regExp);
                    const validCharIdx = _.findIndex(textArr, (eachChar) => charRegExp.test(eachChar));
                    if (validCharIdx !== -1) {
                        maskText.push(textArr[validCharIdx]);
                        _.pullAt(textArr, _.range(0, validCharIdx + 1));
                    } else {
                        return false;
                    }
                }

                const nextChar = textRules[idx + 1];
                return !!textArr.length || !!(nextChar && nextChar.isTemplate);
            });

            return maskText;
        }

        isComplete(maskText) {
            const textRules = this.getTextRules();
            return maskText !== '' && !_.findLast(textRules, (charRule, idx) => {
                if (charRule.isTemplate) {
                    return false;
                } else {
                    const charRegExp = new RegExp(charRule.regExp);
                    return !charRegExp.test(maskText[idx]);
                }
            });
        }

        isEmpty(maskText) {
            const textRules = this.getTextRules();
            return !_.find(textRules, (charRule, idx) => {
                if (charRule.isTemplate) {
                    return false;
                } else {
                    const charRegExp = new RegExp(charRule.regExp);
                    return charRegExp.test(maskText[idx]);
                }
            });
        }

        createGuideMask() {
            return _.map(this.getTextRules(), (charRule) => charRule.guideChar);
        }

        setCursorPos = (inputEl, cursorPos) => {
            inputEl.selectionStart = cursorPos;
            inputEl.selectionEnd = cursorPos;
        }

        getEventData = (e) => {
            const {
                target: inputEl,
                isPaste,
                isKeyDown
            } = e;
            const {
                selectionStart
            } = inputEl;

            return {
                value: inputEl.value,
                selectionStart,
                isPaste: isPaste === true,
                isKeyDown: isKeyDown === true
            };
        }

        _onPaste = (e) => {
            this.dataBeforeChange = this.getEventData({
                ...e,
                isPaste: true
            });
            this.props.onPaste && this.props.onPaste(e);
        }

        _onKeyDown = (e) => {
            this.dataBeforeChange = this.getEventData(e);
            this.props.onKeyDown && this.props.onKeyDown(e);
        }

        _onChange = (e) => {
            this.dataAfterChange = this.getEventData(e);
            const {
                cusorPos,
                maskText
            } = this.getUpdatedData(this.dataBeforeChange, this.dataAfterChange);

            const isUncontrolled = _.isUndefined(this.props.value);
            if (isUncontrolled) {
                e.target.value = maskText;
                this.setCursorPos(e.target, cusorPos);
            }

            this.props.onChange && this.props.onChange(e, {
                cusorPos,
                maskText,
                setCursorPos: this.setCursorPos,
                isComplete: this.isComplete(maskText),
                isEmpty: this.isEmpty(maskText)
            });
        }

        _onFocus = (e) => {
            this.props.onFocus && this.props.onFocus(e);
            // Caculate cursor position
            const textRules = this.getTextRules();
            const currValue = e.target.value;
            const lastValidIdx = _.findLastIndex(textRules, (charRule, idx) => {
                if (charRule.isTemplate || !currValue[idx]) {
                    return false;
                } else {
                    const charRegExp = new RegExp(charRule.regExp);
                    return charRegExp.test(currValue[idx]);
                }
            });
            const editIdx = lastValidIdx === -1 ? 0 : _.findIndex(
                textRules,
                (charRule) => !charRule.isTemplate,
                lastValidIdx + 1
            );
            const cusorPos = editIdx === -1 ? lastValidIdx : editIdx;
            // Set cursor position
            const inputEl = e.target;
            setTimeout(() => this.setCursorPos(inputEl, cusorPos));
        }

        render() {
            const {
                placeholder,
                onChange,
                onFocus,
                onKeyDown,
                onPaste,
                ...restProps
            } = this.props;

            return (
                <TextField
                    {...restProps}
                    placeholder={placeholder || this.createGuideMask().join('')}
                    onChange={this._onChange}
                    onFocus={this._onFocus}
                    onKeyDown={this._onKeyDown}
                    onPaste={this._onPaste}
                />
            );
        }
    }
};

