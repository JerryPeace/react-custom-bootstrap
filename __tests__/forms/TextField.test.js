import React from 'react';
import { mount, shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import TextField from 'forms/TextField';

jest.useFakeTimers();

describe('TextField', () => {
    let textField;

    describe('with default settings', () => {
        test('should render corrently', () => {
            textField = mount(<TextField defaultValue='test' />);
            expect(textField).toMatchSnapshot();
        });
    });

    describe('test getValue function', () => {
        test('should get same value after input given value', () => {
            const testStr = 'test';
            textField = mount(<TextField value={testStr} />).instance();
            expect(textField.getValue()).toEqual(testStr);
        });
    });

    describe('test validator function', () => {
        test('should get correct dom after invalid', () => {
            textField = mount(
                <TextField validator={(me, value) => {
                    return 'this is invalid';
                }} />
            );
            textField.find('input').simulate('change');
            jest.runAllTimers();
            textField.update();
            expect(textField).toMatchSnapshot();
        });
    });

    describe('test isValid function', () => {
        test('should get true after valid', () => {
            const testStr = 'test';
            textField = mount(
                <TextField value={testStr} validator={(me, value) => {
                    return value === testStr;
                }} />
            ).instance();
            expect(textField.isValid()).toBeTruthy();
        });

        test('should get false after invalid', () => {
            const testStr = 'test';
            textField = mount(
                <TextField value={testStr} validator={(me, value) => {
                    return value === 'invalid test';
                }} />
            ).instance();
            expect(textField.isValid()).toBeFalsy();
        });
    });

    describe('test validate function', () => {
        test('should get true after valid', () => {
            const testStr = 'test';
            textField = mount(
                <TextField value={testStr} validator={(me, value) => {
                    return value === testStr;
                }} />
            ).instance();
            expect(textField.validate().result).toBeTruthy();
        });

        test('should get false after invalid', () => {
            const testStr = 'test';
            textField = mount(
                <TextField value={testStr} validator={(me, value) => {
                    return value === 'invalid test';
                }} />
            ).instance();
            expect(textField.validate().result).toBeFalsy();
        });
    });

    describe('test default validator function', () => {
        test('should get true after default validator called', () => {
            textField = mount(<TextField />).instance();
            expect(textField.isValid()).toBeTruthy();
        });
    });

    describe('test mark and clear invalid function', () => {
        test('should get correct helptext dom after function called', () => {
            textField = mount(<TextField />);
            textField.instance().markInvalidText('test');
            textField.update();
            expect(textField).toMatchSnapshot();
            textField.instance().clearInvalidText();
            textField.update();
            expect(textField).toMatchSnapshot();
        });
    });

    describe('test helptext prop', () => {
        test('should render helptext corrently', () => {
            textField = mount(<TextField helptext='test' />);
            expect(textField).toMatchSnapshot();
        });
    });

    describe('test _onChange function', () => {
        test('prop onChange should be called', () => {
            const testValue = 'test';
            let value;
            textField = mount(
                <TextField onChange={(e) => {
                    value = 'test';
                }} />
            ).instance();
            textField._onChange();
            expect(testValue === value).toBeTruthy();
        });

        test('prop onValueChange should be called', () => {
            const testValue = 'test';
            let value;
            textField = mount(
                <TextField onValueChange={(e) => {
                    value = 'test';
                }} />
            ).instance();
            textField._onChange();
            expect(testValue === value).toBeTruthy();
        });
    });
});
