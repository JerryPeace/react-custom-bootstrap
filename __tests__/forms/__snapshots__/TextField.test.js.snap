// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField test helptext prop should render helptext corrently 1`] = `
<Textfield
  bsClass="input"
  delayValidateTime={500}
  disabled={false}
  helpText=""
  helpTextClassName="has-error"
  helptext="test"
  inputType="text"
  labelAlign="left"
  labelClassName="col-xs-2"
  multiline={false}
  standalone={true}
  validator={[Function]}
  wrapperClassName="col-xs-10"
>
  <LayoutHelper
    FieldBase={
      <input
        className=""
        disabled={false}
        helptext="test"
        onChange={[Function]}
        type="text"
      />
    }
    bsClass="input"
    delayValidateTime={500}
    disabled={false}
    getInputEl={[Function]}
    helpText={null}
    helpTextClassName="has-error"
    helptext="test"
    inputType="text"
    labelAlign="left"
    labelClassName="col-xs-2"
    multiline={false}
    singleCheckboxRadio={false}
    standalone={true}
    validator={[Function]}
    wrapperClassName="col-xs-10"
  >
    <div
      className=""
    >
      <input
        className="form-control"
        disabled={false}
        helptext="test"
        onChange={[Function]}
        type="text"
      />
    </div>
  </LayoutHelper>
</Textfield>
`;

exports[`TextField test mark and clear invalid function should get correct helptext dom after function called 1`] = `
<Textfield
  bsClass="input"
  delayValidateTime={500}
  disabled={false}
  helpText=""
  helpTextClassName="has-error"
  inputType="text"
  labelAlign="left"
  labelClassName="col-xs-2"
  multiline={false}
  standalone={true}
  validator={[Function]}
  wrapperClassName="col-xs-10"
>
  <LayoutHelper
    FieldBase={
      <input
        className=""
        disabled={false}
        onChange={[Function]}
        type="text"
      />
    }
    bsClass="input"
    delayValidateTime={500}
    disabled={false}
    getInputEl={[Function]}
    helpText="test"
    helpTextClassName="has-error"
    inputType="text"
    labelAlign="left"
    labelClassName="col-xs-2"
    multiline={false}
    singleCheckboxRadio={false}
    standalone={true}
    validator={[Function]}
    wrapperClassName="col-xs-10"
  >
    <div
      className="has-error"
    >
      <input
        className="form-control"
        disabled={false}
        onChange={[Function]}
        type="text"
      />
      <HelpText
        bsClass="help-block"
        text="test"
      >
        <div
          className=""
        >
          <span
            className="msg help-block"
          >
            test
          </span>
        </div>
      </HelpText>
    </div>
  </LayoutHelper>
</Textfield>
`;

exports[`TextField test mark and clear invalid function should get correct helptext dom after function called 2`] = `
<Textfield
  bsClass="input"
  delayValidateTime={500}
  disabled={false}
  helpText=""
  helpTextClassName="has-error"
  inputType="text"
  labelAlign="left"
  labelClassName="col-xs-2"
  multiline={false}
  standalone={true}
  validator={[Function]}
  wrapperClassName="col-xs-10"
>
  <LayoutHelper
    FieldBase={
      <input
        className=""
        disabled={false}
        onChange={[Function]}
        type="text"
      />
    }
    bsClass="input"
    delayValidateTime={500}
    disabled={false}
    getInputEl={[Function]}
    helpText={null}
    helpTextClassName="has-error"
    inputType="text"
    labelAlign="left"
    labelClassName="col-xs-2"
    multiline={false}
    singleCheckboxRadio={false}
    standalone={true}
    validator={[Function]}
    wrapperClassName="col-xs-10"
  >
    <div
      className=""
    >
      <input
        className="form-control"
        disabled={false}
        onChange={[Function]}
        type="text"
      />
    </div>
  </LayoutHelper>
</Textfield>
`;

exports[`TextField test validator function should get correct dom after invalid 1`] = `
<Textfield
  bsClass="input"
  delayValidateTime={500}
  disabled={false}
  helpText=""
  helpTextClassName="has-error"
  inputType="text"
  labelAlign="left"
  labelClassName="col-xs-2"
  multiline={false}
  standalone={true}
  validator={[Function]}
  wrapperClassName="col-xs-10"
>
  <LayoutHelper
    FieldBase={
      <input
        className=""
        disabled={false}
        onChange={[Function]}
        type="text"
      />
    }
    bsClass="input"
    delayValidateTime={500}
    disabled={false}
    getInputEl={[Function]}
    helpText="this is invalid"
    helpTextClassName="has-error"
    inputType="text"
    labelAlign="left"
    labelClassName="col-xs-2"
    multiline={false}
    singleCheckboxRadio={false}
    standalone={true}
    validator={[Function]}
    wrapperClassName="col-xs-10"
  >
    <div
      className="has-error"
    >
      <input
        className="form-control"
        disabled={false}
        onChange={[Function]}
        type="text"
      />
      <HelpText
        bsClass="help-block"
        text="this is invalid"
      >
        <div
          className=""
        >
          <span
            className="msg help-block"
          >
            this is invalid
          </span>
        </div>
      </HelpText>
    </div>
  </LayoutHelper>
</Textfield>
`;

exports[`TextField with default settings should render corrently 1`] = `
<Textfield
  bsClass="input"
  defaultValue="test"
  delayValidateTime={500}
  disabled={false}
  helpText=""
  helpTextClassName="has-error"
  inputType="text"
  labelAlign="left"
  labelClassName="col-xs-2"
  multiline={false}
  standalone={true}
  validator={[Function]}
  wrapperClassName="col-xs-10"
>
  <LayoutHelper
    FieldBase={
      <input
        className=""
        defaultValue="test"
        disabled={false}
        onChange={[Function]}
        type="text"
      />
    }
    bsClass="input"
    defaultValue="test"
    delayValidateTime={500}
    disabled={false}
    getInputEl={[Function]}
    helpText={null}
    helpTextClassName="has-error"
    inputType="text"
    labelAlign="left"
    labelClassName="col-xs-2"
    multiline={false}
    singleCheckboxRadio={false}
    standalone={true}
    validator={[Function]}
    wrapperClassName="col-xs-10"
  >
    <div
      className=""
    >
      <input
        className="form-control"
        defaultValue="test"
        disabled={false}
        onChange={[Function]}
        type="text"
      />
    </div>
  </LayoutHelper>
</Textfield>
`;
